config-linux.md

```diff
@@ -466,19 +466,37 @@ The following parameters can be specified to set up the controller:
 The following parameters can be specified for the container:
 
 * **`l3CacheSchema`** *(string, OPTIONAL)* - specifies the schema for L3 cache id and capacity bitmask (CBM).
-    If `l3CacheSchema` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
+* **`memBandwidth`** *(array of objects, OPTIONAL)* - an array of memory bandwidth (b/w) percentage per L3 cache id.
+    Each entry has the following structure:
+    * **`id`** *(uint32, REQUIRED)* - L3 cache id
+    * **`value`** *(uint32, REQUIRED)* - memory bandwidth (b/w) percentage
+
+    If `l3CacheSchema` or `memBandwidth` is set, runtimes MUST write the value to the `schemata` file in the `<container-id>` directory discussed in `intelRdt`.
 
-    If `l3CacheSchema` is not set, runtimes MUST NOT write to `schemata` files in any `resctrl` pseudo-filesystems.
+    If neither `l3CacheSchema` nor `memBandwidth` is set, runtimes MUST NOT write to `schemata` files in any `resctrl` pseudo-filesystems.
```

Comment:
I don't think these two set/unset lines are quite right.  The old `l3CacheSchema` requirements were sufficient because the `l3CacheSchema` values were [expected to contain the `L3:` prefix][1].  However, dd23a4c9 only mentions the `MB:` prefix and related formatting the a non-normative Go comment.

If we go the structured route, I (which is where you seem to be taking `memBandwidth`, I think we need to make the `MB:<cache_id0>=bandwidth0;<cache_id1>=bandwidth1;...` formatting a normative part of the spec.  We'd also want to define the interactions between them (e.g. “the formatted lines MAY be written to `schemata` in any order, either with a single write or a series of writes”).  And I think we'd want to structure `l3CacheSchema` (as `l3CacheSchema2?`), both for consistency, and to avoid the ambiguity of:

```json
{
  "l3CacheSchema": "MB:0=50;1=50",
  "memBandwith": [{"id": 0, "value": 25}, {"id": 1, "value": 75}]
}
```


If we go the unstructured route (where `l3CacheSchema` already is), then we can just have an opaque array of strings.  Or even [a single string][2]:

```
"l3CacheSchema": "L3:0=3;1=c\nMB:0=50;1=50"
```

because [JSON allows escaped newlines][3].  At that point, the `l3CacheSchema` property name would be stale so it would be nice to rename it to something more generic (`schemata`?) in v2.  But I think it already does what you need here.

[1]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/config-linux.md#L471
[2]: https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/Documentation/x86/intel_rdt_ui.txt?h=v4.13.8#n197
[3]: https://tools.ietf.org/html/rfc7159#section-7

Reply:
@wking 
> I don't think these two set/unset lines are quite right. The old l3CacheSchema requirements were sufficient because the l3CacheSchema values were expected to contain the L3: prefix. However, dd23a4c only mentions the MB: prefix and related formatting the a non-normative Go comment.

Actually, CAT and MBA features are orthogonal, each one doesn't depend on the other. We could see (1) either "L3 cache" or "memory bandwidth" schema available or (2) both "L3 cache" and "memory bandwidth" schemata are available with specific hardware and kernel configuration.

Reply:
@wking 
> If we go the structured route, (which is where you seem to be taking memBandwidth), I think we need to make the MB:<cache_id0>=bandwidth0;<cache_id1>=bandwidth1;... formatting a normative part of the spec. ...

> If we go the unstructured route (where l3CacheSchema already is), then we can just have an opaque array of strings. Or even a single string: ...

Originally I proposed unstructured JSON format (opaque string: e.g., `"memBwSchema": "MB:0=20;1=70"`) for memory bandwidth schema in https://github.com/opencontainers/runc/issues/1596. 

@cyphar prefers structured JSON format (array of object: e.g., `"memBandwith": [{"id": 0, "value": 20}, {"id": 1, "value": 70}]`). @cyphar's justification in https://github.com/opencontainers/runc/issues/1596#issuecomment-330477399 makes sense (easy for spec validation, use interactively, easy to extend and etc.). In this PR I chose the structured JSON format.

The unstructured (opaque string) format also has some advantages: https://github.com/opencontainers/runc/issues/1596#issuecomment-330328687. It will be easier to conjunct L3 cache and memory bandwidth schema into a single string, and write once to schemata file. I'd like to hear other reviewers' opinions and suggestions.

For L3 cache schema, it is already used opaque string format (e.g., `"l3CacheSchema": "L3:0=7f0;1=1f"`). Anyway, I have plan to unify the formats for L3 cache and memory bandwidth. 

If we choose the structured JSON format, how about this mapping from the schemata file content to JSON config?
```
cat /sys/fs/resctrl/container_id/schemata
L3:0=7f0;1=1f
MB:0=20;1=70
```

```
"linux": {
    "intelRdt": {
        "l3Cache": [{"id": 0, "value": 7f0}, {"id": 1, "value": 1f}],
        "memBandwidth": [{"id": 0, "value": 20}, {"id": 1, "value": 70}]
    }
}
```

Reply:
> The unstructured (opaque string) format also has some advantages…

I agree that there are tradeoffs between structured data and isolating users from the underlying kernel API vs. unstructured data that provides more direct access to the underlying kernel API.  And I don't really care if we go with a structured approach or an unstructured approach.  What doesn't make sense to me is going with an unstructured approach for `l3CacheSchema` and then going for a structured approach in `memBandwidth`.  It seems like the “we should isolate config authors from the details of the kernel API” (or not) would be a decision that applied equally to everything under `intelRdt`.

> Anyway, I have plan to unify the formats for L3 cache and memory bandwidth.

I don't think the spec is a good place to play with the config format, because now that we've cut 1.0.0 with the existing `l3CacheSchema`, we need to continue to support it until this spec hits v2.  If we decided to go with [your unified, structured approach][1] (or [the unified, unstructured `schemata` I'd floated earlier][2], or some other approach), we'd need to continue to support the deprecated `l3CacheSchema` throughout the 1.x spec lifetime and define how the deprecated property interacted with the new properties.  But we certainly don't want to stamp a 1.1 with something that gets deprecated in 1.2, because then we have to support both the 1.0 `l3CacheSchema` and the 1.1 `otherDeprecatedThing` until 2.0.  So if you have plans for something unified, I think they belong in this PR, and not in follow-up work.

[1]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r145560769
[2]: https://github.com/opencontainers/runtime-spec/pull/932#discussion_r145528966

Reply:
@wking 
@cyphar @crosbymichael 
/cc @vishh @mrunalp @vbatts @dqminh @philips @tianon @hqhq

> I don't think the spec is a good place to play with the config format, because now that we've cut 1.0.0 with the existing l3CacheSchema, we need to continue to support it until this spec hits v2. If we decided to go with your unified, structured approach (or the unified, unstructured schemata I'd floated earlier, or some other approach), we'd need to continue to support the deprecated l3CacheSchema throughout the 1.x spec lifetime and define how the deprecated property interacted with the new properties. But we certainly don't want to stamp a 1.1 with something that gets deprecated in 1.2, because then we have to support both the 1.0 l3CacheSchema and the 1.1 otherDeprecatedThing until 2.0. So if you have plans for something unified, I think they belong in this PR, and not in follow-up work.

I understood the consistency and compatibility requirement in runtime-spec. This is my plan:
1. Firstly, I will address "L3 cache" and "memory bandwidth" with `unified` formats in this single PR. 
2. To support existed "l3CacheSchema" throughout 1.x spec lifetime, and to avoid confusion of deprecated property, how about using similar "unified unstructured schemata" format for memory bandwidth in this PR and throughout 1.x runtime-spec and runc ?
3. If we have requirement to change all Intel RDT resources into "structured schemata" in spec 2.0 or newer version, I could open a new PR to slightly rework on appropriate time slot in the phase of spec 2.0.

Please help review and comment. Thank you!

Reply:
> This is my plan…

The plan sounds reasonable to me.  For the 1.x structure, we probably want at least SHOULDs for the `L3:` and `MB:` prefixes and internal newlines, and clear requirements for how we handle cases where those SHOULDs are violated.  MUSTs for those requirements would be nice, and we can do that for `memBandwidth`, but it may be too late to add new MUSTs to `l3CacheSchema`.

And there's also no reason we couldn't specify the 2.x structure now, as long as we document that the `*Schema` property overrides the analogous 2.x property when the 1.x property is set.  In that approach, the plan would be to have this PR deprecate `l3CacheSchema` and add `l3Cache` and `memBandwidth`.  We'd remove `l3CacheSchema` in 2.0.